### DISCUSSION TITLE
Get ready for the final round

### DISCUSSION THREAD

**Discussion by vimal-parakhiya** [tark-ignite/codegenie-2023#7] [Show and tell]
In final round, you will be asked to solve the problem mimicking real world scenario. You will have to apply **Object Oriented principles, 
Data Structures and Algorithmic skills** to effectively solve the problem. 

In order to get ready for the final battle, you may want to start practicing by solving following problem. Once you are done, post your github link (with the solution) in comments section of this page.

### Parking Management System
You have been asked to design the parking management system of a shopping mall. It has N floors reserved for parking. On each floor, there are X parking slots. Out of these, few parking slots are reserved for senior citizens, i.e. the ones which are closer to the entrance.

Each floor would be numbered using alphabets A to Z. For example, if there are 5 floors reserved for parking, they will be numbered as A, B, C, D and E.

Parking slot on each floor will use following conventions, 
FLOOR-SLOT

For example, if there are 10 parkings on each floor, the parking slots on first floor will be numbered as, A-1, A-2, A-3, ..., A-10. Similarly, parking slots on 2nd floor will be numebred as B-1, B-2, B-3,..., B-10

**You have to implement a system, which allocates a parking slot when a new car arrives and computes charge to be paid when it leaves the shopping mall. Additionally, you have to also generate a report with details of parking slot and list of cars, which used that slot. Please see below description for clarity.**

Following are the initial inputs to the program:

1. **No of floors:** An integer, representing number of floors. It will be between 1 and 26, so they can be numbered using alphabets A-Z
2. **No of parking slots on each floor:** An integer representing total parking slots on a floor. This will be between 1 and 100
3. **Parking slots reserved for senior citizens:** A string representing the reserved parkings. For example `A-1 A-10 B-2 B-5 C-1 C-8  D-2 D-4 E-5 E-10`. Each parking slot is separated by space and indicates the exact slot on each floor, reserved for senior citizens.

For example, the initial inputs will look like as follows, indicating there are 5 floors, 10 parking slots on each floor and reserved slots (as a string) on each floor.

```
1. 5
2. 10
3. A-1 A-10 B-2 B-5 C-1 C-8 D-2 D-4 E-5 E-10
```

As a next, the program will accept series of inputs in following format. All these inputs (including above) will be entered using command line/console.

`Input: CHECKIN GJ03DF-2708 08:15am NR`

This indicates the car with number GJ03DF-2708 arrived at 8:15am. **NR** indicates the car does not have any Senior Citizen passengers and so it should use **non-resevred parking slot**. In this case, the program should allocate first available parking slot, but exlcuding the ones, reserved for senior citizens. So, output will be `A-2`

`Output: A-2` (because A-1 is reserved for senior citizen)

`Next Input: CHECKIN GJT-3293 09:00am R`
`Output: A-1` (**R indicates that the reserved parking slot** should be allocated to this car as it has Senior Citizen passengers, and hence A-1 is allocated)

After few more inputs.

`Next Input: CHECKIN GJ05GH-2808 11:00am NR`
`Output: PARKING FULL` (As all parking slots are full)
**Notes:** The reserved slots can be used by senior citizens only.

`Next Input: CHECKOUT GJ03DF-2708 11:45am`
`Output: 80` (The first car arrived at 8:15, leaves after 3.5 hours. So that parking slot is vacated and program displays the charge to be paid)

The charge will be calculated as as follows:

```
<= 2 hours: 50 Rs
>2 hours and <=4 hours: 80 Rs
>4 hours: 100 Rs
```

`Next Input: CHECKOUT GJT-3293 12:00pm`
`Output:  80` (Second car leaves after 3 hours)

`Next Input: CHECKIN GJ09DF-9687 12:15pm NR`
`Output: A-2` (A-2 slot is vacant at this moment, and so it should be allocated to this car, having non-senior citizen passenger)

`Next Input: CHECKOUT GJ09DF-9687 12:45pm`
`Output:  50` (Car left within 30 mins)

`Next Input: GENERATE REPORT`

It will generate and display report in the following format. Output should be sorted by parking slot number, followed by checkin time, as shown below. In the report, include only those cars, which left the mall, i.e. the ones for which CHECK IN and CHECK OUT time are recorded in the system.

```
PARKING SLOT, CAR NO, CHECK IN TIME, CHECK OUT TIME, CHARGES, CATEGORY
A-1, GJT-3293, 09:00am, 12:00pm 80, R
A-2, GJ03DF-2708, 08:15am, 11:45am, 80, NR
A-2, GJ09DF-9687, 12:15pm, 12:45pm, 50, NR
```

**Comment by DeepPatel25** [comment]
**Parking Management System Solution using Java.**
https://github.com/DeepPatel25/CodeGenie_2023/blob/main/Solution/ParkingManagementSystem.java

**Reply by vimal-parakhiya** [reply]
@DeepPatel25 pretty quick! Our team will review this and get back to you.

**Reply by vimal-parakhiya** [reply]
@DeepPatel25 few observations,

1. All inputs should be read from console, including number of floors, parking slots on each floor as well as string representing list of reserved slots. In final round also, it is expected that all inputs are read from console, so that your solution is generic enough and works for different test data
2. Apart from ParkingManagementSystem and Car class, you may also want to define few additional classes like ParkingSlot and ParkingFloor (to have the class and object model, reflecting real world). It will also allow you to reduce the code complexity by distributing logic across relevant classes, e.g. ParkingSlot class can have a flag/logic to determine if it is reserved, if it is allotted etc.
3. Checkout functionality seems to be broken, as it results into exception for following input. On similar lines, you may want to review and fix the report generation functionality as well.

```
5
10 
A-1 A-10 B-2 B-5 C-1 C-8 D-2 D-4 E-5 E-10

Output: 

CHECKIN GJ03DF-2708 08:15am NR
A-2 
CHECKIN GJT-3293 09:00am R
A-1 
CHECKIN GJ05GH-2808 11:00am NR
A-3 
CHECKOUT GJ03DF-2708 11:45am
Exception in thread "main" java.time.format.DateTimeParseException: Text '11:45am' could not be parsed.
```

**Reply by DeepPatel25** [reply]
Thank you sir for reviewing my code.
I will modify it as per your observations.

**Comment by JayDabhi240** [comment]
Parking Management System solution in C++.
https://github.com/Dabhi-Jay/ParkingManagementSystem

**Reply by vimal-parakhiya** [reply]
@Dabhi-Jay our team is reviewing the solution and so we will keep you posted!

**Reply by vimal-parakhiya** [reply]
@Dabhi-Jay solution looks good and mostly correct! Few observations,
1. You may want to define few additional classes like ParkingSlot and ParkingFloor (to have the class and object model, reflecting real world). It will also allow you to reduce the code complexity by distributing logic across relevant classes, e.g. ParkingSlot class can have a flag/logic to determine if it is reserved, if it is allotted etc.
2. You may want to simplify the logic for converting time to minutes, may be two step process could help, i.e. convert it into 24 hour format and then use `H*60 + M` or have two separate functions to extract just hour of the day (returning number between 0 to 23) and minutes. Post that, you can just do `H*60 + M`
3. During checkout, it iterates over all cars/number plates. Instead, you may want to use data structures like Map, to quickly search the car or parking slot by car number. Additionally, current implementation may not work, if user/car visits the mall multiple times in a day, since the loop `breaks` on first match
4. You may want to use proper access modifiers for the fields defined in a class along with required getters/setters and constructors (as needed).

**Reply by JayDabhi240** [reply]
Got it!
Thanks for reviewing my code,I will work to optimize my code as suggested.

**Comment by Saurav-Sutaria** [comment]
Here's my solution:
https://github.com/Saurav-Sutaria/CodeGenie2023_practice

**Reply by vimal-parakhiya** [reply]
@Saurav-Sutaria our team is reviewing the solution and so we will keep you posted!

**Reply by vimal-parakhiya** [reply]
@Saurav-Sutaria good work! Solution is mostly correct!

1. Code is nicely segregated using smaller readable functions. This is good. Easy to follow as well as it'll be easy for you to enhance to accommodate new requirements.
2. As mentioned in above threads, you may also want to try out by creating other classes like ParkingFloor and ParkingSlot to mimic the real world.
3. You may want to use appropriate access modifiers for the fields, defined in the class
4. Looks like, you missed to print "Parking Charges" during checkout, so you may want to correct it. Additionally, you may also want to handle the scenario, wherein there are no reserved slots (in a mall).

**Reply by Saurav-Sutaria** [reply]
Thank you sir for your feedback, I will modify my code as per your suggestions.

**Comment by aminlodhiya** [comment]
Here is my Java Solution!
https://github.com/Mahmadamin08/CodeGenie_2023/blob/main/Final%20Practice/ParkingManagementSystem.java

**Reply by vimal-parakhiya** [reply]
@Mahmadamin08 solution is concise and readable! Usage of meaningful names (for variables and methods) is helpful.

As mentioned above, you may also want to try creating classes to represent ParkingFloor and ParkingSlot.

Additionally, Checkout function seems to be broken. It is not displaying "Parking Charges" as well as computation seems to be incorrect, so you may want to review and fix it.

For example, it appears to be calculating incorrect charges for car:  `GJ05GH-2808 (11:00am to 04:00pm)`

```
5 
10 
A-1 A-10 B-2 B-5 C-1 C-8 D-2 D-4 E-5 E-10
 
Actual Output: 
 
CHECKIN GJ03DF-2708 08:15am NR 
A-2
CHECKIN GJT-3293 09:00am R
A-1
CHECKIN GJ05GH-2808 11:00am NR
A-3
CHECKOUT GJ03DF-2708 11:45am
CHECKOUT GJT-3293 12:00am
CHECKOUT GJ05GH-2808 04:00pm
GENERATE REPORT                            
PARKING SLOT, CAR NO, CHECK IN TIME, CHECK OUT TIME, CHARGES, CATEGORY
A-1 , GJT-3293 , 09:00am , 12:00am , 80 , R
A-2 , GJ03DF-2708 , 08:15am , 11:45am , 80 , NR
A-3 , GJ05GH-2808 , 11:00am , 04:00pm , 50 , NR
 
Expected Output: 
 
CHECKIN GJ03DF-2708 08:15am NR 
A-2 
CHECKIN GJT-3293 09:00am R 
A-1 
CHECKIN GJ05GH-2808 11:00am NR 
A-3 
CHECKOUT GJ03DF-2708 11:45am 
80 
CHECKOUT GJT-3293 12:00am 
80 
CHECKOUT GJ05GH-2808 04:00pm 
100 
GENERATE REPORT 
PARKING SLOT, CAR NO, CHECK IN TIME, CHECK OUT TIME, CHARGES, CATEGORY 
A-1, GJT-3293, 09:00am, 12:00am, 80, R 
A-2, GJ03DF-2708, 08:15am, 11:45am, 80, NR 
A-3, GJ05GH-2808, 11:00am, 04:00pm, 100, NR 
```

**Reply by aminlodhiya** [reply]
Thank you!!  for your feedback, 
I have Updated my Code. kindly review and let me know if still have any mistake.

**Reply by vimal-parakhiya** [reply]
@Mahmadamin08 looks like you may want to fix the logic to read input data from console.

It appears to be failing on inputs such as,
```
Input:
5
10
A-1

Output:
Invalid Details...
```

If the user enters 10 (no of slots) as the integer input and then hits Enter key, the `nextInt()` method will read the 10, but the newline character `(\n) ` that follows the 10 will still be in the input stream. 

When the `nextLine()` method is called, it will read this newline character and return an empty string. 

This may cause unexpected behavior in your program. 

You can add `sc.nextLine()` after `sc.nextInt()` or you can use `sc.nextLine()` to read the input (including integer values) and then convert the `string` to `int` using `Integer.parseInt(strInput.trim())`

Please note that in final round (tomorrow), you will have to read data from command line, using same techniques. So, you may want to be aware about such issues and techniques to resolve them, to save the time during final battle!

**Reply by aminlodhiya** [reply]
Yes! Got it,
but in this Program first give input 5 and then give 1space and then give input 10 and give 1space and give input Reserved String and hit enter ..
then the program will run perfectly..
is it valid to separate inputs by Space ?

**Reply by aminlodhiya** [reply]
I have Updated my Code!!

**Reply by vimal-parakhiya** [reply]
No, you will have to read inputs as mentioned in the problem description, i.e. new line separated. 

In final round also, some of the inputs will be new line separated, whereas some of the inputs could be space separated. It will be mentioned in the problem description. 

As we have to assess code of multiple students, we have to stick to the input format, as mentioned in the problem description.

**Reply by aminlodhiya** [reply]
Ok,
Thank You!

**Comment by Dev-Bhayani-001** [comment]
Parking management system solution in java
https://github.com/Dev-Bhayani-001/Parking_management_system

**Reply by vimal-parakhiya** [reply]
@Dev-Bhayani-001 nice work, you may want to look into following and clean-up few things.

1. Number of reserved slots are hard-coded, you may want to change it to ArrayList or dynamic array
2. You may want to remove the debug output, which prints parking matrix
3. As mentioned above, defining few other classes like ParkingFloor and ParkingSlot would help in distributing the logic (and hence code complexity).
4. Instead of `in/out`, the variables can have meaningful names.

**Comment by ManD-2113** [comment]
Solution of this problem statement
https://github.com/ManD-2113/Parking-Mannagement-System

**Comment by DeepPatel25** [comment]
Respected @vimal-parakhiya,
At the final round of CodeGenie-2023 I tried very well but fail to clear milestone 1 because of 1 bug. I tried at my home and fix that bug and passed all testcase of Milestone 1 so I am request to you to please check my code and give your valuable response for my efforts.

Here is the link of my solution :- 
https://github.com/DeepPatel25/tark-codegenie-2023-final

**Reply by vimal-parakhiya** [reply]
Sure @DeepPatel25, we will check this out and keep you posted!

**Reply by vimal-parakhiya** [reply]
@DeepPatel25 our team has reviewed this and following are few suggestions.

Overall solution looks mostly correct, however, you may want to test it against test data mentioned in [this](https://bit.ly/codegenie-2023-all-tests) link, which includes couple more tests (apart from default set of tests shared earlier).

1. Looks like, if there are multiple booking requests for the same date, it creates multiple objects of type `DayWiseTrainbooking`, instead of having only one object per date
2. Instead of temp, temp1, storeHashmap, storeData etc, you may use meaningful variable names, as it was bit hard to follow/review the code (by others)
3. While parsing the coach input detail, you may use (SL, 3A, 2A and 1A) as keys in the hashmap, namely `bookingSlot`, so that processing a booking request would be easier and you do not have to map SL against S, 3A against B and likewise. Additionally, `bookingSlot` can be renamed as `bookedTicketsByCoachType` or `bookedTicketsByTravellingClass`
4. `trainCheck(inputBookingRequest[0], inputBookingRequest[1], storeData)` is called twice. You may want to reduce it to only single call to avoid iterating same list of trains multiple times.

But overall, I'm glad that you walked extra mile and tried to solve the problem (and did not give up!)

**Reply by DeepPatel25** [reply]
Thank you @vimal-parakhiya for your observations.

**Comment by Saurav-Sutaria** [comment]
Hello @vimal-parakhiya Sir,
Here's my code for the milestone 1, it is passing all the testcases of milestone 1.
Will you please verify it once.

https://github.com/Saurav-Sutaria/tark-codegenie-2023-final

**Reply by vimal-parakhiya** [reply]
Sure @Saurav-Sutaria  we will review it and keep you posted.

**Reply by vimal-parakhiya** [reply]
@Saurav-Sutaria solution looks good. You may want to test it using additional test data as mentioned [here](https://bit.ly/codegenie-2023-all-tests).

1. Code is nicely structured using meaningful classes. Methods are short and easy to follow. Naming conventions are also good.
2. At few places, e.g. `dateFree` can be renamed to meaningful name, e.g. `isSeatReservedForDate`
3. Instead of `vector<string>`, you may want to use` set<string>` or something equivalent, so that we do not have to iterate over the list, while checking existence of a date in a set/collection.
4. While creating a coach object, you may also want to store the coachType (like SL, 1A, 2A, 3A) so that while booking a ticket, it will be easier to check if a coach is suitable for a given booking request or not.

Overall, the solution is very good and your approach is very much in the right direction of solving the entire problem, and so I will encourage you to solve Milestone 2 as well. Above test link captures additional test cases for Milestone 1 and Milestone 2 both.

**Reply by Saurav-Sutaria** [reply]
Thank you sir for your review, I will modify my code as per your suggestions and will post it again once I solve the milestone 2.

**Reply by Saurav-Sutaria** [reply]
Sir, I have completed the milestone 2. All the level given in the testcase file are cleared.
Here's my [Code](https://github.com/Saurav-Sutaria/tark-codegenie-2023-final).

Will you please verify it once.

**Reply by vimal-parakhiya** [reply]
Great! Our team will surely review and get back to you!

**Reply by vimal-parakhiya** [reply]
@Saurav-Sutaria  solution looks great! Using vector<Ticket> within Seat class looks correct. Also the logic for checking if a seat is occupied for a given route, looks good too. Overall, the correct usage of class and object model along with meaningful functions, make the code readable! 

Keep it up!

As a next (and just for fun), would you like to enhance it further to support **Waitlisting** and **Cancellations**, i.e.

1. If user submits a booking request and if enough seats are not available, then ticket is booked with waitlist status. For example, assuming there are only 3 seats available in SL class and if user would like to book a ticket with 5 seats (passengers), it will successfully book tickets for 3 passengers, but 2 passengers will have waitlist status. For this ticket the PNR status will look like below,
`100000045, 2023-03-25, 22717, Rajkot, Pune, 4500, S1-1 S1-2 S1-3 WL-1 WL-2`

2. Subsequently, if any user cancels the ticket, the passengers with waitlist status will be confirmed. Passenger with WL-1 will be confirmed first (and a seat is allocated), then passenger with WL-2 status will be confirmed, and so on.
For cancellation, you may accept input in following format, i.e. user having ticket with PNR 100000028 wants to cancel bookings for 3 passengers.
`CANCEL 100000028 3`

Let us know your thoughts or questions, if any!

**Reply by Saurav-Sutaria** [reply]
Sure Sir, I will enhance it further.
Sir if 2 tickets are in waitlist and 3 tickets are confirmed then we have to calculate the price for 3 tickets and if the waitlisted tickets are confirmed in future then we have to modify the fare again, right?

**Reply by vimal-parakhiya** [reply]
We will calculate fare for all passengers, including passengers in wait list status. So, in above example, we should compute fare for all 5 passengers.

However, if someone cancels the ticket, we should recompute the fair for that booking/PNR.

**Reply by Saurav-Sutaria** [reply]
Okay,
For cancellation do have to free the seats from the beginning (like cancel S1-1 first then S1-2) or we have to free it from end.
And if a person cancel all his booking then the PNR number associated with the ticket will get freed or it will still refer to the same ticket with zero booking?
And can we cancel the waiting tickets or the cancel command is only the booked tickets?

**Reply by Saurav-Sutaria** [reply]
For each user the waiting list will start from beginning, (like WL-1 WL-2)
or the waiting list number will continue from where the previous user's left
User 1 : WL-1 WL-2
User 2 : WL-3 WL-4 WL-5
?

**Reply by vimal-parakhiya** [reply]
> Okay, For cancellation do have to free the seats from the beginning (like cancel S1-1 first then S1-2) or we have to free it from end. And if a person cancel all his booking then the PNR number associated with the ticket will get freed or it will still refer to the same ticket with zero booking? And can we cancel the waiting tickets or the cancel command is only the booked tickets?

1. For cancellations, we will free-up only those seats, which are cancelled and the same seats will be allocated to passengeres with waitlist status. For example, user has booked a ticket with seats as `S1-35 S1-36 S1-37 S1-38 S1-39`. Now, if user cancels 2 seats from this booking, then `S1-38 S1-39` will be freed-up and same seats will be allocated to the passengers with waitlist status as `WL-1 and WL-2`
2. If a person cancels the entire ticket (all seats in that booking request), the PNR number still remains valid. However, if we get details by that PNR, it will return total fare as 0, i.e. the ticket with zero bookings/no passengers
3. Yes, we can cancel confirmed seats/tickets as well as the waitlisted tickets

**Reply by vimal-parakhiya** [reply]
> For each user the waiting list will start from beginning, (like WL-1 WL-2) or the waiting list number will continue from where the previous user's left User 1 : WL-1 WL-2 User 2 : WL-3 WL-4 WL-5 ?

Wait list numbers will continue from where the previous user/list is left, so as you have mentioned, it will be,
```
User 1 : WL-1 WL-2
User 2 : WL-3 WL-4 WL-5
```

However, if someone cancels 3 confirmed seats (S1-40, S1-41 and S1-42), then passengers with status WL-1, WL-2 and WL-3 are marked as confirmed, and so the new status will look like as follows, i.e. now there are only 2 passengers with waitlist status and so they are marked as WL-1 and WL-2
```
User 1 : S1-40 S1-41
User 2 : S1-42 WL-1 WL-2
```

**Reply by Saurav-Sutaria** [reply]
> 3. Yes, we can cancel confirmed seats/tickets as well as the waitlisted tickets.

What's the command to cancel the waitlisted tickets?
Can we give it like : 
CANCEL Waiting 100000028 2

**Reply by vimal-parakhiya** [reply]
The command should be same as earlier, i.e.

`CANCEL 100000028 3`

Assuming there are total 5 seats booked against this PNR `S1-70 S1-71 S1-72 WL-1 WL-2` and if user cancels 3 seats (for this PNR), then 2 waitlisted tickets and 1 confirmed ticket/seat will be canceled. So, new seat status for this PNR will be `S1-70 S1-71`

Another scenario, assuming there are total 5 seats booked against PNR `100000035`, i.e. `S1-72 WL-1 WL-2 WL-3 WL-4` and if user cancels 3 waitlisted tickets against this PNR `(CANCEL 100000035 3)`, the new status will look like `S1-72 WL-1`

**Reply by Saurav-Sutaria** [reply]
ok sir, got it.

**Reply by Saurav-Sutaria** [reply]
Sir, I have added the Cancellation and waiting list feature.
Will you please verify my [Code.](https://github.com/Saurav-Sutaria/tark-codegenie-2023-final)

(The feature of updating the waiting list numbers after the user cancels some tickets is pending.)

**Reply by vimal-parakhiya** [reply]
Cool @Saurav-Sutaria , we will review this and meanwhile keep us posted once you solve the remaining problem statement.

**Reply by Saurav-Sutaria** [reply]
Sir, I have completed the task of updating the waiting list.
Please review it.

**Reply by Saurav-Sutaria** [reply]
@vimal-parakhiya Sir, Do you have any testcases for this cancellation and waiting list's task?

**Reply by vimal-parakhiya** [reply]
Hey @Saurav-Sutaria  

Glad to know that you were able to solve it fully!

Sorry, for delayed response. Since last few days, the team is busy in planned project work, however, I've certainly have this in my radar to review your code and provide feedbacks, if any.

We do not have any test cases for this requirement, but you may create on your own, as per the format of test cases shared earlier. If you get a chance to prepare some, please also share it with us so we can use them when review/validate your code.

Happy coding!
